research/workflow-engines/airflow.md

```diff
@@ -0,0 +1,72 @@
+---
+title: Airflow
+authors: lbarcziova
+---
+
+- docs: https://airflow.apache.org/docs/apache-airflow/stable/index.html
+
+  > Apache Airflow is a platform used to programmatically author, schedule, and monitor workflows. It allows you to define complex workflows as directed acyclic graphs (DAGs) and execute them on a scheduled basis or triggered by events.
+
+- A DAG is the core concept in Airflow. It's a collection of tasks and their dependencies, organized as a directed acyclic graph. Example:
+
+```python
+from datetime import datetime
+from airflow import DAG
+from airflow.operators.python_operator import PythonOperator
+
+def build_task():
+    pass
+
+def test_task():
+    pass
+
+def dynamic_workflow():
+    dag = DAG(
+        f'dynamic_workflow',
+        start_date=datetime.now(),
+        schedule_interval=None,
+        catchup=False,
+    )
+
+    build = PythonOperator(
+        task_id=f'build',
+        python_callable=build_task,
+        op_args=[],
+        dag=dag,
+    )
+
+    test = PythonOperator(
+        task_id=f'test',
+        python_callable=test_task,
+        op_args=[],
+        dag=dag,
+    )
+
+    # Set task dependencies
+    build >> test
+
+    return dag
+
+dynamic_workflow()
+
+```
+
+- Airflow uses schedulers and executors to manage task execution. The scheduler determines when to execute tasks, while the executor defines how tasks are run. Airflow supports various executors, including SequentialExecutor, LocalExecutor, and CeleryExecutor.
+- allows dynamic pipelines generation
+- provides UI ![UI](https://airflow.apache.org/docs/apache-airflow/stable/_images/dags.png)
+- from `Why not` in docs:
+  > Airflow™ was built for finite batch workflows. While the CLI and REST API do allow triggering workflows, Airflow was not built for infinitely running event-based workflows. Airflow is not a streaming solution. However, a streaming system such as Apache Kafka is often seen working together with Apache Airflow. Kafka can be used for ingestion and processing in real-time, event data is written to a storage location, and Airflow periodically starts a workflow processing a batch of data.
+
+## CeleryExecutor
+
+Details: https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/celery.html
+
+- it is also possible to use Flower UI
```

Comment:
This looks interesting, have you had time to investigate how would this work? (What is the benefit of using Celery as an executor and if/how we can transition from pure Celery to Airflow+Celery?)

Reply:
for us, the clear benefit would be using a tool that we already know, the configuration, deployment around it etc.

but as for the transition, this way, Celery would be used for executing the tasks, but the Airflow schedulers would handle the logic of scheduling them, so the way tasks are handled would be changed.

there are multiple types of executors that can be used, see above, for the remote execution of tasks, besides Celery, one can also use KubernetesExecutor, CeleryKubernetesExecutor, etc. (I will add this here so that it is more explicit)

Reply:
Maybe a `KubernetesExecutor` can be used for/intead-of the Sandcastle.

Reply:
I'm sure that over the years, a solution better than sandcastle has to be out there. The idea of sandcastle is simple and we're not the only one that need that.

Anyway, I found https://github.com/confidential-containers/confidential-containers/blob/main/quickstart.md

which look promising, but under the hood, it's kata containers ( = VMs) that use [TEE](https://en.wikipedia.org/wiki/Trusted_execution_environment)s: this solution is far more expensive than sandcastle

Red Hat's docs: https://access.redhat.com/documentation/en-us/openshift_sandboxed_containers/1.4/html-single/openshift_sandboxed_containers_user_guide/index#deploying-sandboxed-container-workloads-peer-pods